be/src/runtime/batch_write/txn_state_cache.cpp

```diff
@@ -0,0 +1,467 @@
+                          << ", status: " << result.status();
+        // fast fail if there is failure between FE and BE
+        if (_num_poll_failure >= config::merge_commit_txn_state_poll_max_fail_times) {
+            _transit_txn_state(TTransactionStatus::UNKNOWN,
+                               fmt::format("poll txn state failure exceeds max times {}, last error: {}",
+                                           _num_poll_failure, result.status().to_string(false)),
+                               false);
+        }
+    } else {
+        TRACE_BATCH_WRITE << "handler poll success, txn_id: " << _txn_id << ", " << result.value();
+        _num_poll_failure = 0;
+        _transit_txn_state(result.value().txn_status, result.value().reason, false);
+    }
+    // stop polling if reach the finished state or there is no subscriber
+    if (_is_finished_txn_state()) {
+        _cv.notify_all();
+        return false;
+    } else {
+        return _num_subscriber > 0;
+    }
+}
+
+void TxnStateHandler::subscribe(bool& trigger_poll) {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    _num_subscriber++;
+    // trigger polling if this is the first subscriber and not in finished state
+    trigger_poll = _num_subscriber == 1 && !_is_finished_txn_state();
+}
+
+void TxnStateHandler::unsubscribe() {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    _num_subscriber--;
+}
+
+StatusOr<TxnState> TxnStateHandler::wait_finished_state(const std::string& subscriber_name, int64_t timeout_us) {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    if (_is_finished_txn_state()) {
+        return _txn_state;
+    }
+    if (_stopped) {
+        return Status::ServiceUnavailable("Transaction state handler is stopped");
+    }
+    _num_waiting_finished_state++;
+    DeferOp defer([&] { _num_waiting_finished_state--; });
+
+    int64_t left_timeout_us = timeout_us;
+    while (left_timeout_us > 0) {
+        TRACE_BATCH_WRITE << "start to wait state, subscriber name: " << subscriber_name << ", txn_id: " << _txn_id
+                          << ", timeout_us: " << left_timeout_us;
+        auto start_us = MonotonicMicros();
+        int ret = _cv.wait_for(lock, left_timeout_us);
+        int64_t elapsed_us = MonotonicMicros() - start_us;
+        TRACE_BATCH_WRITE << "finish to wait state, subscriber name: " << subscriber_name << ", txn_id: " << _txn_id
+                          << ", elapsed: " << elapsed_us << " us, txn_status: " << to_string(_txn_state.txn_status)
+                          << ", reason: " << _txn_state.reason << ", stopped: " << _stopped;
+        if (_is_finished_txn_state()) {
+            return _txn_state;
+        } else if (_stopped) {
+            return Status::ServiceUnavailable("Transaction state handler is stopped");
+        } else if (ret == ETIMEDOUT) {
+            break;
+        }
+        left_timeout_us = std::max((int64_t)0, left_timeout_us - elapsed_us);
```

Comment:
when is `elapsed_us` smaller than `left_timeout_us`? except the cv wait is successful.

Reply:
it serves as a protective mechanism to avoid inaccuracies in time couting.